### PR TITLE
[Core]support more kwargs for `--defer`

### PR DESCRIPTION
**Pull request by MyronFanQiu**
---
to support 
```
cached_put(cmd, client.create_or_update, rule_group, resource_group_name=resource_group_name, firewall_policy_name=firewall_policy_name, rule_group_name=rule_group_name)
```
This checklist is used to make sure that common guidelines for a pull request are followed.

- [ ] The PR has modified HISTORY.rst describing any customer-facing, functional changes. Note that this does not include changes only to help content. (see [Modifying change log](https://github.com/Azure/azure-cli/tree/master/doc/authoring_command_modules#modify-change-log)).

- [ ] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

### CODE REVIEW

**Comment by tjprescott** [comment]
Is there a linked issue that makes it clear what wasn't working and why this will fix it?

**Comment by MyronFanQiu** [comment]
@tjprescott there is a limitation.
```
        if len(args) > 2:=
            raise CLIError('expected 2 args, got {}: {}'.format(len(args), args))
        if len(copy_kwargs) > 1:	
            raise CLIError('expected 1 kwarg, got {}: {}'.format(len(copy_kwargs), copy_kwargs))
```

I cannot use `cached_put` like
```
cached_put(cmd, client.create_or_update, rule_group, resource_group_name=resource_group_name, firewall_policy_name=firewall_policy_name, rule_group_name=rule_group_name)
```
or
```
cached_put(cmd, client.create_or_update, rule_group, resource_group_name, firewall_policy_name, rule_group_name)
```

I need three parameters(resource_group_name, firewall_policy_name and rule_group_name) to store/get a cache. Is there another way to do it without changing the code?

**Comment by MyronFanQiu** [comment]
I deleted the check and make the concatenation of `copy_kwargs` as the file name.

**Comment by tjprescott** [comment]
> @tjprescott there is a limitation.
> 
> ```
>         if len(args) > 2:=
>             raise CLIError('expected 2 args, got {}: {}'.format(len(args), args))
>         if len(copy_kwargs) > 1:	
>             raise CLIError('expected 1 kwarg, got {}: {}'.format(len(copy_kwargs), copy_kwargs))
> ```
> 
> I cannot use `cached_put` like
> 
> ```
> cached_put(cmd, client.create_or_update, rule_group, resource_group_name=resource_group_name, firewall_policy_name=firewall_policy_name, rule_group_name=rule_group_name)
> ```
> 
> or
> 
> ```
> cached_put(cmd, client.create_or_update, rule_group, resource_group_name, firewall_policy_name, rule_group_name)
> ```
> 
> I need three parameters(resource_group_name, firewall_policy_name and rule_group_name) to store/get a cache. Is there another way to do it without changing the code?

The way the cache was intended is that you do not cache child objects, in this case a rule group. You would cache just the firewall policy. In the same way, you don't cache a subnet when doing a subnet create--you cache the vnet that the subnet belongs to. That is why it ensures that only one "name" parameter is submitted. Caching child objects instead of just top level resources could lead to confusing inconsistencies. For example, if a rule group was cached and then a PUT was done on a firewall policy, it would not know about the cached rule group and would submit whatever was in the firewall policy (cached or not).

**Comment by yonzhan** [comment]
@jiasli please take a look.

File: src/azure-cli-core/azure/cli/core/commands/__init__.py [hunk 1]
```diff
@@ -132,13 +132,14 @@ def path(self, args, kwargs):
...
-            raise CLIError('expected 1 kwarg, got {}: {}'.format(len(copy_kwargs), copy_kwargs))
 
+        directory_appendix = None
         try:
             resource_name = args[-1]
         except IndexError:
-            resource_name = list(copy_kwargs.values())[0]
+            resource_list = list(copy_kwargs.values())
+            resource_name = resource_list.pop()
+            directory_appendix = resource_list
```
Comment (jiasli) [inline src/azure-cli-core/azure/cli/core/commands/__init__.py]: `_model_name` for sub-resource doesn't seem to be included in the path.

**Comment by yonzhan** [comment]
How is the status of this PR?

**Comment by MyronFanQiu** [comment]
close this one first and add it into backlog

**Pull request closed without merging**